alibi_detect/utils/loading.py

```diff
@@ -0,0 +1,625 @@
+# TODO - Need to modularise torch and tensorflow imports and use. e.g. has_tensorflow and has_pytorch etc
+from alibi_detect.utils.registry import registry
+from alibi_detect.utils.schemas import SupportedModels  # type: ignore[attr-defined]
+from alibi_detect.utils.validate import validate_config
+from alibi_detect.utils.tensorflow._loading import load_model as load_model_tf, \
+    prep_model_and_emb as prep_model_and_emb_tf, load_kernel_config as load_kernel_config_tf, \
+    load_optimizer as load_optimizer_tf, load_embedding as load_embedding_tf, load_detector_legacy, Detectors
+#  from alibi_detect.utils.pytorch._loading import set_device
+import tensorflow as tf  # TODO - this is currently only required for FIELDS_TO_DTYPE conversion. Remove in future.
+from transformers import AutoTokenizer
+import numpy as np
+from typing import Union, Optional, Callable, Any
+from copy import deepcopy
+from functools import partial
+from pathlib import Path
+import dill
+import os
+import toml
+from importlib import import_module
+import logging
+
+logger = logging.getLogger(__name__)
+
+REQUIRES_BACKEND = [
+    'ClassifierDrift',
+    'ClassifierUncertaintyDrift',
+    'LearnedKernelDrift',
+    'LSDDDrift',
+    'MMDDrift',
+    'RegressorUncertaintyDrift',
+    'SpotTheDiffDrift'
+]
+
+# Fields to resolve in resolve_cfg ("resolve" meaning either load local artefact or resolve @registry, conversion to
+# tuple, np.ndarray and np.dtype are dealt with separately).
+FIELDS_TO_RESOLVE = [
+    ['preprocess_fn'],
+    ['preprocess_fn', 'src'],
+    ['preprocess_fn', 'model'],
+    ['preprocess_fn', 'embedding'],
+    ['preprocess_fn', 'tokenizer'],
+    ['x_ref'],
+    ['model'],
+    ['optimizer'],
+    ['reg_loss_fn'],
+    ['kernel'],
+    ['dataset'],
+    ['kernel', 'src'],
+    ['kernel', 'proj'],
+    ['kernel', 'kernel_a', 'src'],
+    ['kernel', 'kernel_b', 'src'],
+    ['initial_diffs']
+]
+
+# Directories to amend before resolving config (fields to prepend config file dir to)
+DIR_FIELDS = [
+    ['preprocess_fn'],
+    ['preprocess_fn', 'src'],
+    ['preprocess_fn', 'model', 'src'],
+    ['preprocess_fn', 'embedding', 'src'],
+    ['preprocess_fn', 'tokenizer', 'src'],
+    ['x_ref'],
+    ['model', 'src'],
+    ['kernel'],
+    ['kernel', 'src'],
+    ['optimizer'],
+    ['reg_loss_fn'],
+    ['kernel', 'proj', 'src'],
+    ['kernel', 'kernel_a', 'src'],
+    ['kernel', 'kernel_b', 'src'],
+    ['initial_diffs']
+]
+
+# Fields to convert from list to tuple in resolve_cfg
+FIELDS_TO_TUPLE = [
+    ['detector', 'kwargs', 'input_shape']
+]
+
+# Fields to convert from list to np.ndarray in resolve_cfg
+FIELDS_TO_ARRAY = [
+    ['sigma'],
+    ['kernel', 'sigma'],
+    ['kernel', 'kernel_a', 'sigma'],
+    ['kernel', 'kernel_b', 'sigma'],
+]
+
+# Fields to convert from str to np.dtype
+FIELDS_TO_DTYPE = [
+    ['preprocess_fn', 'dtype']
+]
```

Comment:
I have yet to read the rest of the logic so perhaps this comment is premature - but I thought `pydantic` should take care of this (e.g. feeding a `list` into a `pydantic` model that expects a `tuple` would coerce into a `tuple`; this can be extended to custom validators so we can also do `list` -> `np.ndarray` coercion).

Reply:
That's a good point, I shall explore this.

Reply:
I'm now using pydantic to coerce to `Tuple` and `np.ndarray` as suggested. I've left `FIELDS_TO_DTYPE` as is for now.

Note: I use the new `NDArray` validator when typing simple ndarray's in the pydantic models (such as for `sigma`). I've left `x_ref` typed with the less strict `np.ndarray` for now, as would need to think about how to extend the `NDArray` class to accept `Any` for this.

Reply:
@jklaise leaving this unresolved so that you can check you are happy with the above.

Reply:
Nice! Looks much cleaner. Left a couple of comments on the `NDArray` type validator too.